File: src/10.graph/25.282.ts
export function addOperators(num: string, target: number): string[] {
    const ans: string[] = [];
    const n = num.length;
    const backtrack = (start: number, pre: number, cur: number, s: string) => {
        if (start === n) {
            if (cur === target) {
                ans.push(s);
            }
            return;
        }

        for (let i = start; i < n; i++) {
            if (i > start && num[start] == "0") {
                break;
            }

            const val = num.slice(start, i + 1);
            const t = Number(val);

            if (start === 0) {
                backtrack(i + 1, t, t, val);
            } else {
                backtrack(i + 1, t, cur + t, s + "+" + val);
                backtrack(i + 1, -t, cur - t, s + "-" + val);

                // * 有优先计算的权利
                backtrack(i + 1, pre * t, cur - pre + pre * t, s + "*" + val);
            }
        }
    };

    backtrack(0, 0, 0, "");

    return ans;
}

File: src/10.graph/25.282.go
package graph

import (
	"strconv"
)

func AddOperators(num string, target int) []string {
	ans := []string{}
	n := len(num)

	var backtrack func(r int, prev int, cur int, s string)

	backtrack = func(r int, prev int, cur int, s string) {
		if r == n {
			if cur == target {
				ans = append(ans, s)
			}
			return
		}

		for i := r; i < n; i++ {
			end := i + 1
			val := num[r:end]
			if len(val) > 1 && val[0] == '0' {
				break
			}

			t, _ := strconv.Atoi(val)

			if r == 0 {
				backtrack(i+1, t, t, val)
			} else {
				backtrack(i+1, t, cur+t, s+"+"+val)
				backtrack(i+1, -t, cur-t, s+"-"+val)
				backtrack(i+1, prev*t, cur-prev+prev*t, s+"*"+val)
			}

		}
	}

	backtrack(0, 0, 0, "")

	return ans
}
